### PR TITLE
unified labels names

### PR DESCRIPTION
**Pull request by rcerven**
#499, creates label class for use across whole osbs

### CODE REVIEW

**Comment by twaugh** [comment]
Relates to: https://github.com/projectatomic/atomic-reactor/pull/606

**Comment by twaugh** [comment]
Let's all be clear on what the interface is supposed to look like before making any more code changes.

**Comment by twaugh** [comment]
My original proposal for it: https://github.com/projectatomic/atomic-reactor/pull/606#issuecomment-270402677

File: tests/test_utils.py [hunk 1]
```diff
@@ -239,3 +239,53 @@ def test_get_instance_token_file_name():
...
 
     assert get_instance_token_file_name('spam') == expected
+
+@pytest.mark.parametrize(('labels', 'fnc', 'expect'), [
+        ({},
+         ("get_label_only", Labels.LABEL_TYPE_COMPONENT),
+         "com.redhat.component"),
+        ({},
+         ("get_label_only", "new_label"),
+         "new_label"),
```
Comment (twaugh) [inline tests/test_utils.py]: No, this test is wrong. "new label" is not a valid key.

File: tests/test_utils.py [hunk 2]
```diff
@@ -239,3 +239,53 @@ def test_get_instance_token_file_name():
...
+        ({},
+         ("get_label_only", ""),
+         ""),
```
Comment (twaugh) [inline tests/test_utils.py]: "" is not a valid key. This test should be removed.

File: tests/test_utils.py [hunk 3]
```diff
@@ -239,3 +239,53 @@ def test_get_instance_token_file_name():
...
+          "Version": "version", "Architecture": "architecture",
+          "Release": "release", "BZComponent": "com.redhat.component",
+          "Authoritative_Registry": "authoritative-source-url"}),
+        ({"Name" : "old",
+          "name" : "new"},
+          ("get_label_and_value", Labels.LABEL_TYPE_NAME),
+          ("name", "new")),
+        ({},
+          ("get_label_and_value", "doest_exist"),
+          KeyError),
```
Comment (twaugh) [inline tests/test_utils.py]: "doest_exist" is not a valid key -- calling this method with some invalid key is not defined behaviour and we should not test it.

File: tests/test_utils.py [hunk 4]
```diff
@@ -239,3 +239,53 @@ def test_get_instance_token_file_name():
...
+        ({"some_label" : "some_value"},
+          ("get_label_and_value", "some_label"),
+          ("some_label", "some_value"))
```
Comment (twaugh) [inline tests/test_utils.py]: "some_label" is not a valid key.

The only valid keys are the ones defined in the Labels class.

File: tests/test_utils.py [hunk 5]
```diff
@@ -239,3 +239,53 @@ def test_get_instance_token_file_name():
...
+    ])
```
Comment (twaugh) [inline tests/test_utils.py]: Missing test case:
```
({},
 ("get_label_and_value", Labels.TYPE_LABEL_NAME),
 KeyError),
```

File: tests/test_utils.py [hunk 6]
```diff
@@ -239,3 +239,47 @@ def test_get_instance_token_file_name():
...
 
     assert get_instance_token_file_name('spam') == expected
+
+@pytest.mark.parametrize(('labels', 'fnc', 'expect'), [
+        ({},
+         ("get_label_only", Labels.LABEL_TYPE_COMPONENT),
+         "com.redhat.component"),
+        ({},
+         ("get_label_only", "doesnt_exist"),
+         KeyError),
```
Comment (twaugh) [inline tests/test_utils.py]: Maybe `Exception`? We're outside the specification of how this function ought to behave, so if it were to raise e.g. `TypeError` that would be perfectly valid behaviour and the test should still pass.

File: tests/test_utils.py (hunk 3 above)
Comment (twaugh) [inline tests/test_utils.py]: If you want to test this, expect `Exception` rather than `KeyError`.

File: tests/test_utils.py [hunk 7]
```diff
@@ -239,3 +239,47 @@ def test_get_instance_token_file_name():
...
+          "Authoritative_Registry": "authoritative-source-url"}),
+        ({"Name" : "old",
+          "name" : "new"},
+          ("get_label_and_value", Labels.LABEL_TYPE_NAME),
+          ("name", "new")),
+        ({},
+          ("get_label_and_value", "doest_exist"),
+          KeyError),
+    ])
+def test_labels(labels, fnc, expect):
```
Comment (twaugh) [inline tests/test_utils.py]: Missing test: get_label_and_value called with a valid label type but for which the Dockerfile provides no label.

Comment (twaugh) [inline tests/test_utils.py]: Should be:
```
({},
 ("get_label_and_value", Labels.LABEL_TYPE_NAME),
 KeyError),
```

File: tests/test_utils.py [hunk 8]
```diff
@@ -239,3 +239,50 @@ def test_get_instance_token_file_name():
...
+          "Version": "version", "Architecture": "architecture",
+          "Release": "release", "BZComponent": "com.redhat.component",
+          "Authoritative_Registry": "authoritative-source-url"}),
+        ({"Name" : "old",
+          "name" : "new"},
+         ("get_label_and_value", Labels.LABEL_TYPE_NAME),
+         ("name", "new")),
+        ({},
+         ("get_label_and_value", Labels.LABEL_TYPE_NAME),
+         Exception),
```
Comment (twaugh) [inline tests/test_utils.py]: We can be more specific here -- we definitely want this to be an instance of `KeyError`.

File: tests/test_utils.py [hunk 9]
```diff
@@ -239,3 +239,50 @@ def test_get_instance_token_file_name():
...
+         KeyError),
+        ({},
+         ("get_label_and_value", "doest_exist"),
+         Exception),
+    ])
+def test_labels(labels, fnc, expect):
+    label = Labels(labels)
+
+    fn, arg = fnc
+    if isinstance(expect, type):
```
Comment (twaugh) [inline tests/test_utils.py]: Should this be `isinstance(expect, Exception)`?

File: osbs/utils.py [hunk 10]
```diff
@@ -358,3 +358,68 @@ def run_command(*popenargs, **kwargs):
             message="Command %s returned %s\n\n%s" % (cmd, retcode, output)
         )
     return output
+
+
+class Labels(object):
```
Comment (twaugh) [inline osbs/utils.py]: Could include a class docstring here, something like:
```
    """
    Provide access to a set of labels which have specific semantics

    The set of labels comes from here:
    https://github.com/projectatomic/ContainerApplicationGenericLabels

    Note that only a subset of label types (those used by OSBS) is supported:

    - LABEL_TYPE_NAME: repository name of the image 
    - LABEL_TYPE_VERSION: version of the image
    - LABEL_TYPE_RELEASE: release number for this version
    - LABEL_TYPE_ARCH: architecture for the image
    - LABEL_TYPE_VENDOR: owner of the image
    - LABEL_TYPE_SOURCE: authoritative location for publishing
    - LABEL_TYPE_COMPONENT: Bugzilla (or other tracker) component
    - LABEL_TYPE_HOST: build host used to create the image
    """
```

File: tests/test_utils.py (hunk 9 above)
Comment (rcerven) [inline tests/test_utils.py]: but I am passing class of exception not instance

Comment (twaugh) [inline tests/test_utils.py]: Ah, right. OK, this is fine as it is.

**Comment by twaugh** [comment]
This looks good!

Seeing as osbs-client only has one place where it looks at labels (OSBS.create_prod_build in osbs/api.py), how about modifying that to use this new class?

**Comment by twaugh** [comment]
Looks good.

Now, before we merge, is everyone happy with the method names? One issue I see is that they have different ways of referring to the name of the label -- `label` in `get_label_only()` vs `name` in `get_name_and_value()` and `get_new_names_by_old()`. Do we want to make that consistent?

**Comment by vrutkovs** [comment]
>Do we want to make that consistent?

I think `label` should be present, as `get_label_name_and_value` (maybe `get_label_items`?) and `get_label_names_by_old` sound better to me

File: osbs/api.py [hunk 11]
```diff
@@ -361,26 +361,15 @@ def create_prod_build(self, git_uri, git_ref,
...
-
         df_parser = utils.get_df_parser(git_uri, git_ref, git_branch=git_branch)
         build_request = self.get_build_request()
-        labels = df_parser.labels
+        labels = utils.Labels(df_parser.labels)
 
-        name_label = get_required_label(labels, ('name', 'Name'))
-        component = get_required_label(labels, ('com.redhat.component', 'BZComponent'))
+        try:
+            _, name_label = labels.get_label_and_value(utils.Labels.LABEL_TYPE_NAME)
```
Comment (vrutkovs) [inline osbs/api.py]: `name_label` here is actually label's value

**Comment by rcerven** [comment]
maybe get_name_only() instead get_label_only() ?

**Comment by rcerven** [comment]
get_label_items() instead get_label_name_and_value() is nice

but as for get_new_names_by_old()
this seems to be nicely describing what it returns

**Comment by lcarva** [comment]
`get_label_items()` is confusing to me, because it implies it'll behave similarly to `dict.items()` which is not the case since it only returns one item, perhaps `get_label_item()` is more accurate? But I still think `get_label_name_and_value()` better describes the method behavior. To keep it consistent I think either we should have `_label_` in both of these methods or omit them, not a mix. For example:

```python
get_name_and_value()
get_name()
```
or
```python
get_label_name_and_value()
get_label_name()
```

**Comment by rcerven** [comment]
this seem nice:
get_name_and_value()
get_name()

**Comment by twaugh** [comment]
There's something to be said about *not* having `label` in the method name -- after all, it's already in the class name.

**Comment by vrutkovs** [comment]
+1 to `get_name` and `get_name_and_value`

**Comment by rcerven** [comment]
seems like all like : 
get_name_and_value()
get_name()

so what about : 
get_new_names_by_old()

**Comment by vrutkovs** [comment]
@rcerven, how about `get_preferred` or `get_preferred_for_old`?

**Comment by rcerven** [comment]
preferred was before used as "best option" so not always 'new'

**Comment by vrutkovs** [comment]
`_PREFERRED_LABELS` is sorted like that, I guess its safe to assume that 'new' == 'preferred'

File: osbs/utils.py [hunk 12]
```diff
@@ -358,3 +358,85 @@ def run_command(*popenargs, **kwargs):
...
+
+    def __init__(self, df_labels):
+        """
+        Create a new Labels object
+        providing access to actual newest labels as well as old ones
+        """
+        self._df_labels = df_labels
+        self._label_values = {}
+        for label_type in Labels.LABEL_NAMES:
+            for lbl_name in Labels.LABEL_NAMES[label_type]:
```
Comment (lcarva) [inline osbs/utils.py]: Could be nicer to use `items()` here:

```python
for label_type, label_names in Labels.LABEL_NAME.items():
  for lable_name in label_names:
...
```

File: osbs/utils.py [hunk 13]
```diff
@@ -358,3 +358,85 @@ def run_command(*popenargs, **kwargs):
...
+
+    @staticmethod
+    def get_new_names_by_old():
+        """Return dictionary, new label name indexed by old label name."""
+        newdict = {}
+
+        for label_type, label_names in Labels.LABEL_NAMES.items():
+            for oldname in label_names[1:]:
+                newdict[oldname] = Labels.LABEL_NAMES[label_type][0]
+        return newdict
```
Comment (vrutkovs) [inline osbs/utils.py]: This method returns a dict, unlike others. 

Why not just return a new name instead?
Where would this method be used?

**Comment by twaugh** [comment]
If we can't agree on how `get_new_names_by_old()` should behave or should be named, how about dropping this method until we actually want to use it?

File: osbs/utils.py (hunk 13 above)
Comment (rcerven) [inline osbs/utils.py]: this is suppose to replace 
in pre_add_labels_in_df.py
DEFAULT_ALIASES dictionary

**Comment by rcerven** [comment]
squashed

**Pull request merged**